File: app/src/main/java/org/eyeseetea/malariacare/factories/AuthenticationFactory.java
package org.eyeseetea.malariacare.factories;

import android.content.Context;

import org.eyeseetea.malariacare.data.boundaries.ICredentialsDataSource;
import org.eyeseetea.malariacare.data.database.datasources.CredentialsLocalDataSource;
import org.eyeseetea.malariacare.data.repositories.AuthenticationManager;
import org.eyeseetea.malariacare.data.repositories.CredentialsRepository;
import org.eyeseetea.malariacare.domain.boundary.executors.IAsyncExecutor;
import org.eyeseetea.malariacare.domain.boundary.executors.IMainExecutor;
import org.eyeseetea.malariacare.domain.boundary.repositories.IAuthenticationManager;
import org.eyeseetea.malariacare.domain.boundary.repositories.ICredentialsRepository;
import org.eyeseetea.malariacare.domain.usecase.GetCredentialsUseCase;
import org.eyeseetea.malariacare.domain.usecase.LoginUseCase;
import org.eyeseetea.malariacare.domain.usecase.LogoutUseCase;
import org.eyeseetea.malariacare.presentation.executors.AsyncExecutor;
import org.eyeseetea.malariacare.presentation.executors.UIThreadExecutor;

public class AuthenticationFactory {
    private IMainExecutor mMainExecutor = new UIThreadExecutor();
    private IAsyncExecutor mAsyncExecutor = new AsyncExecutor();

    public LoginUseCase getLoginUseCase(Context context) {
        IAuthenticationManager authenticationManager = new AuthenticationManager(context);
        LoginUseCase loginUseCase = new LoginUseCase(authenticationManager, mMainExecutor,
                mAsyncExecutor);
        return loginUseCase;
    }

    public LogoutUseCase getLogoutUseCase(Context context) {
        IAuthenticationManager authenticationManager = new AuthenticationManager(context);
        LogoutUseCase logoutUseCase = new LogoutUseCase(authenticationManager, mMainExecutor,
                mAsyncExecutor);
        return logoutUseCase;
    }

    public GetCredentialsUseCase getLoadCredentialsUseCase(Context context) {
        ICredentialsDataSource credentialsDataSource = new CredentialsLocalDataSource(context);
        ICredentialsRepository credentialsRepository = new CredentialsRepository(
                credentialsDataSource);
        return new GetCredentialsUseCase(mAsyncExecutor, mMainExecutor, credentialsRepository);
    }


}

File: app/src/main/java/org/eyeseetea/malariacare/factories/AuthenticationFactory.kt
package org.eyeseetea.malariacare.factories

import android.content.Context
import org.eyeseetea.malariacare.data.database.datasources.ServerInfoLocalDataSource
import org.eyeseetea.malariacare.data.remote.api.ServerInfoRemoteDataSource
import org.eyeseetea.malariacare.data.repositories.ServerInfoRepository
import org.eyeseetea.malariacare.data.repositories.UserAccountRepository
import org.eyeseetea.malariacare.data.repositories.UserD2ApiRepository
import org.eyeseetea.malariacare.domain.boundary.repositories.IServerRepository
import org.eyeseetea.malariacare.domain.boundary.repositories.IUserAccountRepository
import org.eyeseetea.malariacare.domain.usecase.LoginUseCase
import org.eyeseetea.malariacare.domain.usecase.LogoutUseCase
import org.eyeseetea.malariacare.presentation.executors.AsyncExecutor
import org.eyeseetea.malariacare.presentation.executors.UIThreadExecutor
import org.eyeseetea.malariacare.presentation.executors.WrapperExecutor
import org.eyeseetea.malariacare.presentation.presenters.LoginPresenter

object AuthenticationFactory {
    private val mainExecutor = UIThreadExecutor()
    private val asyncExecutor = AsyncExecutor()

    fun provideLoginPresenter(context: Context): LoginPresenter {
        val executor = WrapperExecutor()
        val getServersUseCase = ServerFactory.provideGetServersUseCase(context)

        return LoginPresenter(executor, getServersUseCase)
    }

    fun provideLoginUseCase(context: Context): LoginUseCase {

        val userAccountRepository = provideUserAccountRepository(context)
        val serverLocalDataSource = ServerInfoLocalDataSource(context)
        val serverRemoteDataSource = ServerInfoRemoteDataSource(context)
        val serverInfoRepository = ServerInfoRepository(
            serverLocalDataSource,
            serverRemoteDataSource
        )
        val userRepository = UserD2ApiRepository()

        val serverRepository: IServerRepository = ServerFactory.provideServerRepository(context)

        return LoginUseCase(
            userAccountRepository,
            serverRepository,
            serverInfoRepository,
            userRepository,
            mainExecutor,
            asyncExecutor
        )
    }

    fun provideLogoutUseCase(context: Context): LogoutUseCase {
        var mUserAccountRepository: IUserAccountRepository = provideUserAccountRepository(context)

        return LogoutUseCase(mUserAccountRepository)
    }

    private fun provideUserAccountRepository(context: Context) =
        UserAccountRepository(context)
}